Source/Renderer/ComputeEngine.js

```diff
@@ -0,0 +1,213 @@
+        return fbo;
+    }
+
+    function createViewportQuadShader(context, fragmentShaderSource) {
+        return ShaderProgram.fromCache({
+            context : context,
+            vertexShaderSource : ViewportQuadVS,
+            fragmentShaderSource : fragmentShaderSource,
+            attributeLocations : viewportQuadAttributeLocations
+        });
+    }
+
+    function createRenderState(width, height) {
+        if ((!defined(renderStateScratch)) ||
+            (renderStateScratch.viewport.width !== width) ||
+            (renderStateScratch.viewport.height !== height)) {
+
+            renderStateScratch = RenderState.fromCache({
+                viewport : new BoundingRectangle(0, 0, width, height)
+            });
+        }
+        return renderStateScratch;
+    }
+
+    ComputeEngine.prototype.execute = function(computeCommand) {
+        //>>includeStart('debug', pragmas.debug);
+        if (!defined(computeCommand)) {
+            throw new DeveloperError('computeCommand is required.');
+        }
+        //>>includeEnd('debug');
+
+        if (defined(computeCommand.preExecutionCallback)) {
+            computeCommand.preExecutionCallback(computeCommand);
+        }
+
+        //>>includeStart('debug', pragmas.debug);
+        if (!defined(computeCommand.fragmentShaderSource) && !defined(computeCommand.shaderProgram)) {
+            throw new DeveloperError('computeCommand.fragmentShaderSource or computeCommand.shaderProgram is required.');
+        }
+
+        if (!defined(computeCommand.outputTexture)) {
+            throw new DeveloperError('computeCommand.outputTexture is required.');
+        }
+        //>>includeEnd('debug');
+
+        var texture = computeCommand.outputTexture;
+        var width = texture.width;
+        var height = texture.height;
+
+        var context = this._context;
+        var vertexArray = defined(computeCommand.vertexArray) ? computeCommand.vertexArray : getViewportQuadVertexArray(this);
+        var shaderProgram = defined(computeCommand.shaderProgram) ? computeCommand.shaderProgram : createViewportQuadShader(context, computeCommand.fragmentShaderSource);
+        var framebuffer = createFramebuffer(context, texture);
+        var renderState = createRenderState(width, height);
+        var uniformMap = defaultValue(computeCommand.uniformMap, defaultValue.EMPTY_OBJECT);
+
+        var clearCommand = clearCommandScratch;
+        clearCommand.framebuffer = framebuffer;
+        clearCommand.renderState = renderState;
+
+        var drawCommand = drawCommandScratch;
+        drawCommand.vertexArray = vertexArray;
+        drawCommand.renderState = renderState;
+        drawCommand.shaderProgram = shaderProgram;
+        drawCommand.uniformMap = uniformMap;
+        drawCommand.framebuffer = framebuffer;
+
+        clearCommand.execute(context);
+        drawCommand.execute(context);
+        framebuffer.destroy();
+
+        if (!computeCommand.persists) {
```

Comment:
I know this will change with pooling, but we need to think through the semantics of `persists` a bit more since, for example, a persistence compute command never empties the shader cache...maybe it is OK for now.